qa/smoke-test-http/src/test/java/org/elasticsearch/http/TestDeprecationHeaderRestAction.java

```diff
@@ -66,18 +69,26 @@
 
     private final Settings settings;
 
-    public TestDeprecationHeaderRestAction(Settings settings, RestController controller) {
+    public TestDeprecationHeaderRestAction(Settings settings) {
         this.settings = settings;
+    }
 
-        controller.registerAsDeprecatedHandler(RestRequest.Method.GET, "/_test_cluster/deprecated_settings", this,
-                                               DEPRECATED_ENDPOINT, deprecationLogger);
+    @Override
+    public List<DeprecatedRoute> deprecatedRoutes() {
+        return singletonList(
```

Comment:
I haven't looked deeply but I wonder if we need both a separate method and a separate class for deprecated routes. Could you return Deprecated routes in the handledRoutes method?

Reply:
I looked at that briefly and decided against doing so to avoid `instanceOf` checks

Reply:
I'm not sure if this is better, it's merely a suggestion for how we could do something like this:

```diff
diff --git a/server/src/main/java/org/elasticsearch/rest/RestController.java b/server/src/main/java/org/elasticsearch/rest/RestController.java
index 6a16007465c..90763fd1336 100644
--- a/server/src/main/java/org/elasticsearch/rest/RestController.java
+++ b/server/src/main/java/org/elasticsearch/rest/RestController.java
@@ -48,6 +48,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Predicate;
 import java.util.function.Supplier;
 import java.util.function.UnaryOperator;
 import java.util.stream.Collectors;
@@ -62,6 +63,7 @@ import static org.elasticsearch.rest.RestStatus.OK;
 public class RestController implements HttpServerTransport.Dispatcher {
 
     private static final Logger logger = LogManager.getLogger(RestController.class);
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(logger);
 
     private final PathTrie<MethodHandlers> handlers = new PathTrie<>(RestUtils.REST_DECODER);
 
@@ -130,11 +132,7 @@ public class RestController implements HttpServerTransport.Dispatcher {
      */
     protected void registerWithDeprecatedHandler(RestRequest.Method method, String path, RestHandler handler,
                                               RestRequest.Method deprecatedMethod, String deprecatedPath,
-                                              DeprecationLogger logger) {
-        // e.g., [POST /_optimize] is deprecated! Use [POST /_forcemerge] instead.
-        final String deprecationMessage =
-            "[" + deprecatedMethod.name() + " " + deprecatedPath + "] is deprecated! Use [" + method.name() + " " + path + "] instead.";
-
+                                              String deprecationMessage, DeprecationLogger logger) {
         registerHandler(method, path, handler);
         registerAsDeprecatedHandler(deprecatedMethod, deprecatedPath, handler, deprecationMessage, logger);
     }
@@ -160,11 +158,12 @@ public class RestController implements HttpServerTransport.Dispatcher {
      * and {@code path} combinations.
      */
     public void registerHandler(final RestHandler restHandler) {
-        restHandler.routes().forEach(route -> registerHandler(route.getMethod(), route.getPath(), restHandler));
-        restHandler.deprecatedRoutes().forEach(route ->
-            registerAsDeprecatedHandler(route.getMethod(), route.getPath(), restHandler, route.getDeprecationMessage(), route.getLogger()));
-        restHandler.replacedRoutes().forEach(route -> registerWithDeprecatedHandler(route.getMethod(), route.getPath(),
-            restHandler, route.getDeprecatedMethod(), route.getDeprecatedPath(), route.getLogger()));
+        final List<RestHandler.Route> routes = restHandler.routes();
+        routes.stream().filter(Predicate.not(RestHandler.Route::isDeprecated)).forEach(route -> registerHandler(route.getMethod(), route.getPath(), restHandler));
+        routes.stream().filter(RestHandler.Route::isDeprecated).forEach(route ->
+            registerAsDeprecatedHandler(route.getMethod(), route.getPath(), restHandler, route.deprecationMessage(), deprecationLogger));
+        routes.stream().filter(RestHandler.Route::isReplacement).forEach(route -> registerWithDeprecatedHandler(route.getMethod(), route.getPath(),
+            restHandler, route.replacedRoute().getMethod(), route.replacedRoute().getPath(), route.replacedRoute().deprecationMessage(), deprecationLogger));
     }
 
     @Override
diff --git a/server/src/main/java/org/elasticsearch/rest/RestHandler.java b/server/src/main/java/org/elasticsearch/rest/RestHandler.java
index ab7b468f757..9638933053d 100644
--- a/server/src/main/java/org/elasticsearch/rest/RestHandler.java
+++ b/server/src/main/java/org/elasticsearch/rest/RestHandler.java
@@ -20,7 +20,6 @@
 package org.elasticsearch.rest;
 
 import org.elasticsearch.client.node.NodeClient;
-import org.elasticsearch.common.logging.DeprecationLogger;
 import org.elasticsearch.common.xcontent.XContent;
 import org.elasticsearch.rest.RestRequest.Method;
 
@@ -72,32 +71,30 @@ public interface RestHandler {
         return Collections.emptyList();
     }
 
-    /**
-     * A list of routes handled by this RestHandler that are deprecated and do not have a direct
-     * replacement. If changing the {@code path} or {@code method} of a route,
-     * use {@link #replacedRoutes()}.
-     */
-    default List<DeprecatedRoute> deprecatedRoutes() {
-        return Collections.emptyList();
-    }
-
-    /**
-     * A list of routes handled by this RestHandler that have had their {@code path} and/or
-     * {@code method} changed. The pre-existing {@code route} will be registered
-     * as deprecated alongside the updated {@code route}.
-     */
-    default List<ReplacedRoute> replacedRoutes() {
-        return Collections.emptyList();
-    }
-
     class Route {
 
         private final String path;
         private final Method method;
+        private final DeprecationInfo deprecationInfo;
+        private final ReplacementInfo replacementInfo;
 
         public Route(Method method, String path) {
+            this(method, path, DeprecationInfo.NON_DEPRECATED_ROUTE, ReplacementInfo.NONE);
+        }
+
+        public Route(Method method, String path, Method deprecatedMethod, String deprecatedPath) {
+            this(
+                method,
+                path,
+                DeprecationInfo.NON_DEPRECATED_ROUTE,
+                ReplacementInfo.replacedRoute(new Route(deprecatedMethod, deprecatedPath, DeprecationInfo.deprecatedRoute("[" + deprecatedMethod.name() + " " + deprecatedPath + "] is deprecated! Use [" + method.name() + " " + path + "] instead."), ReplacementInfo.NONE)));
+        }
+
+        private Route(Method method, String path, DeprecationInfo deprecationInfo, ReplacementInfo replacementInfo) {
             this.path = path;
             this.method = method;
+            this.deprecationInfo = deprecationInfo;
+            this.replacementInfo = replacementInfo;
         }
 
         public String getPath() {
@@ -107,58 +104,88 @@ public interface RestHandler {
         public Method getMethod() {
             return method;
         }
+
+        public boolean isDeprecated() {
+            return deprecationInfo.isDeprecated();
+        }
+
+        public String deprecationMessage() {
+            return deprecationInfo.deprecationMessage();
+        }
+
+        public boolean isReplacement() {
+            return replacementInfo.isReplacement();
+        }
+
+        public Route replacedRoute() {
+            return replacementInfo.replacedRoute();
+        }
+
     }
 
-    /**
-     * Represents an API that has been deprecated and is slated for removal.
-     */
-    class DeprecatedRoute extends Route {
+    class DeprecationInfo {
 
+        private final boolean isDeprecated;
         private final String deprecationMessage;
-        private final DeprecationLogger logger;
 
-        public DeprecatedRoute(Method method, String path, String deprecationMessage, DeprecationLogger logger) {
-            super(method, path);
+        public static final DeprecationInfo NON_DEPRECATED_ROUTE = new DeprecationInfo(false, null);
+
+        public static DeprecationInfo deprecatedRoute(String deprecationMessage) {
+            return new DeprecationInfo(true, deprecationMessage);
+        }
+
+        private DeprecationInfo(boolean isDeprecated, String deprecationMessage) {
+            assert (isDeprecated && deprecationMessage != null) || (isDeprecated == false && deprecationMessage == null);
+            this.isDeprecated = isDeprecated;
             this.deprecationMessage = deprecationMessage;
-            this.logger = logger;
         }
 
-        public String getDeprecationMessage() {
-            return deprecationMessage;
+        public boolean isDeprecated() {
+            return isDeprecated;
         }
 
-        public DeprecationLogger getLogger() {
-            return logger;
+        public String deprecationMessage() {
+            if (isDeprecated) {
+                return deprecationMessage;
+            } else {
+                throw new IllegalStateException();
+            }
         }
+
     }
 
-    /**
-     * Represents an API that has had its {@code path} or {@code method} changed. Holds both the
-     * new and previous {@code path} and {@code method} combination.
-     */
-    class ReplacedRoute extends Route {
+    class ReplacementInfo {
+
+        private final boolean isReplacement;
+        private final Route replacedRoute;
 
-        private final String deprecatedPath;
-        private final Method deprecatedMethod;
-        private final DeprecationLogger logger;
+        public static final ReplacementInfo NONE = new ReplacementInfo(false, null);
 
-        public ReplacedRoute(Method method, String path, Method deprecatedMethod, String deprecatedPath, DeprecationLogger logger) {
-            super(method, path);
-            this.deprecatedMethod = deprecatedMethod;
-            this.deprecatedPath = deprecatedPath;
-            this.logger = logger;
+        public static ReplacementInfo replacedRoute(Route replacedRoute) {
+            return new ReplacementInfo(true, replacedRoute);
         }
 
-        public String getDeprecatedPath() {
-            return deprecatedPath;
+        private ReplacementInfo(boolean isReplacement, Route replacedRoute) {
+            assert (isReplacement && replacedRoute != null) || (isReplacement == false && replacedRoute == null);
+            if (replacedRoute.isDeprecated() == false) {
+                throw new IllegalArgumentException();
+            }
+            this.isReplacement = isReplacement;
+            this.replacedRoute = replacedRoute;
         }
 
-        public Method getDeprecatedMethod() {
-            return deprecatedMethod;
+        public boolean isReplacement() {
+            return isReplacement;
         }
 
-        public DeprecationLogger getLogger() {
-            return logger;
+        public Route replacedRoute() {
+            if (isReplacement) {
+                return replacedRoute;
+            } else {
+                throw new IllegalStateException();
+            }
         }
+
     }
+
 }
diff --git a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestDeleteExpiredDataAction.java b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestDeleteExpiredDataAction.java
index 11ca99d9d3a..faa96d23755 100644
--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestDeleteExpiredDataAction.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestDeleteExpiredDataAction.java
@@ -5,9 +5,7 @@
  */
 package org.elasticsearch.xpack.ml.rest;
 
-import org.apache.logging.log4j.LogManager;
 import org.elasticsearch.client.node.NodeClient;
-import org.elasticsearch.common.logging.DeprecationLogger;
 import org.elasticsearch.rest.BaseRestHandler;
 import org.elasticsearch.rest.RestRequest;
 import org.elasticsearch.rest.action.RestToXContentListener;
@@ -22,21 +20,10 @@ import static org.elasticsearch.rest.RestRequest.Method.DELETE;
 
 public class RestDeleteExpiredDataAction extends BaseRestHandler {
 
-    private static final DeprecationLogger deprecationLogger =
-        new DeprecationLogger(LogManager.getLogger(RestDeleteExpiredDataAction.class));
-
     @Override
     public List<Route> routes() {
-        return Collections.emptyList();
-    }
-
-    @Override
-    public List<ReplacedRoute> replacedRoutes() {
-        // TODO: remove deprecated endpoint in 8.0.0
         return Collections.singletonList(
-            new ReplacedRoute(DELETE, MachineLearning.BASE_PATH + "_delete_expired_data",
-                DELETE, MachineLearning.PRE_V7_BASE_PATH + "_delete_expired_data",
-                deprecationLogger)
+            new Route(DELETE, MachineLearning.BASE_PATH + "_delete_expired_data", DELETE, MachineLearning.PRE_V7_BASE_PATH + "_delete_expired_data")
         );
     }
```

Reply:
The more I think about it, the more I think we should keep it how you have it for now. If we find reasons to combine this into a single pattern in the future, the above gives us a way forward on how we do that. Let's leave it out of this change though, this is already a huge improvement over where we were, and no need to delay getting it in any further when it's not clear what the benefits are.